File: LeetCode/Medium/CPP/424LongestRepeatingCharacterReplacement.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int characterReplacement(string s, int k)
    {
        vector<int> alpha(26, 0);

        int i = 0, j = 0;
        int maxcount = 0;
        int ans = 0;

        while (j < s.size())
        {
            alpha[s[j] - 'A']++;
            for (int h = 0; h < 26; h++)
                maxcount = max(maxcount, alpha[h]);
            while (j - i + 1 - maxcount > k)
            {
                alpha[s[i] - 'A']--;
                i++;
            }
            ans = max(ans, j - i + 1);
            j++;
        }
        return (ans);
    }
};

int main(void)
{
    Solution s;

    cout << s.characterReplacement("AABABBA", 1) << endl;
    return (0);
}

File: LeetCode/Medium/PYTHON/424LongestRepeatingCharacterReplacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        i,j,maxcount, ans = 0,0,0,0
        alpha = {}
        for j in range(len(s)):
            alpha[s[j]] = 1 + alpha.get(s[j], 0)
            maxcount = max(alpha.values())
            while (j - i + 1- maxcount > k):
                alpha[s[i]] = alpha.get(s[i]) - 1
                i += 1
            ans = max(ans, j - i + 1)
            j += 1
        return ans


s = Solution()

ans = s.characterReplacement("ABAB", 2)

print(ans)